Allow commit_version_bump to find the Settings.bundle in the project

jdee opened the pull request:
Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes:

- [x ] Run `bundle exec rspec` from the subdirectory of each tool you modified. Alternatively, run `rake test_all` from the root directory.
 Some tests are failing in master and continue to fail here. This action actually has no tests. I may add some.
- [x ] Run `bundle exec rubocop -a` to ensure the code style is valid
- [x ] Read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x ] We currently don't accept new actions, please publish a plugin instead, more information in [Plugins.md](https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Plugins.md)

Before submitting a pull request, we appreciate if you create an issue first to discuss the change :+1:


Please see issue #5758.

This change allows the `commit_version_bump` action to infer the location of the Settings.bundle from the project. It also adds a new optional :settings_file parameter that allows the user to specify a plist file in the Settings.bundle other than Root.plist.

Commit:
Allow commit_version_bump to find the Settings.bundle in the project

jdee commented:
This is also related to #6482.

Commit:
rubocop correction

jdee commented:
There are probably not too many cases where an app might include multiple settings plist files in a version bump, but it doesn't seem too far-fetched to think that someone might want to update their `Root.plist` as well as an `About.plist` and so on.

A common Ruby idiom is to allow different types to be passed to an argument, e.g.

```ruby
commit_version_bump xcodeproj: "MyProject.xcodeproj", settings: true # Root.plist
commit_version_bump xcodeproj: "MyProject.xcodeproj", settings: "About.plist"
commit_version_bump xcodeproj: "MyProject.xcodeproj", settings: [ "Root.plist", "About.plist" ]
```

This might be simpler than adding multiple new arguments like `settings_file:` (currently in this branch) and `settings_files:`. But I'm not sure if the `ConfigItem` can support this. I suppose I could check. Either way, it wouldn't be hard to support including multiple files. 

Please let me know if further work is needed for this PR to be merged. Thanks.

ohwutup commented:
Hey @jdee ! Thanks for the PR, apologies for the delay in reviewing! 

This looks good... I think it would be great to have everything in one option (`ConfigItem`). I don't think you should have a problem allowing it to be multiple types by not specifying the type and then just inspecting the type at runtime.

Let me know if you have any questions!

ohwutup commented on fastlane/lib/fastlane/actions/commit_version_bump.rb:
```diff
@@ -66,6 +66,14 @@ def self.run(params)
         expected_changed_files << info_plist_files
+
+        if params[:settings]
+          file = params[:settings_file]
```
Would be great to combine these.

ohwutup commented on fastlane/lib/fastlane/actions/commit_version_bump.rb:
```diff
@@ -147,6 +151,12 @@ def self.available_options
+                                       description: "A plist file in Settings.bundle to include if :settings is true",
+                                       optional: true,
+                                       default_value: "Root.plist",
+                                       type: String),
```
Just skip the type specification here.

KrauseFx commented on fastlane/lib/fastlane/actions/commit_version_bump.rb:
```diff
@@ -192,6 +202,20 @@ def self.example_code
+          raise "No Settings.bundle in project" if settings_bundle.nil?
+
+          project_parent = File.dirname xcodeproj_path
+          File.join project_parent, settings_bundle.path, settings_file_name
```
Please surround this with `(` and `)`

jdee committed:
Support different types for settings: argument

jdee committed:
Added unit tests and a bit of refactoring

jdee committed:
Specify an exception type

jdee commented:
Hi @ohwutup. Sorry for the delay. I had an app to get out. Thanks for the feedback. I think this is much nicer now. In fact these are all now equivalent:

```ruby
commit_version_bump xcodeproj: "MyApp.xcodeproj", settings: true
commit_version_bump xcodeproj: "MyApp.xcodeproj", settings: "Root.plist"
commit_version_bump xcodeproj: "MyApp.xcodeproj", settings: ["Root.plist"]
```

I also added some tests for the methods I added. Please let me know if there's anything left to be done. Thanks.

ohwutup commented:
Thanks for making those changes @jdee ! If you don't mind, can we just change one more little thing about this? I think that `settings` is a bit of an ambiguous name for the option. Can we change that to be something a bit more descriptive?

jdee commented:
Hi @ohwutup. I'd be happy to change that. Any suggestions? Maybe `include_settings`, `settings_plist(s)`, `settings_file(s)`, `include_settings_file(s)`, `include_settings_plists(s)`?

Also, what if someone is using the current syntax. I know it was broken, but in a couple of apps at my last job, I was using `settings: true, force: true` and I just put the Settings.bundle at the top of the repo. Maybe the action could report an error if the old argument is supplied and say, "Please use the new syntax." Or both could be supported for now. Or just forget about it and change it?

jdee committed:
Change config argument to include_settings: to match the env. var.

jdee committed:
Also use the new parameter

ohwutup commented:
Oh! My bad @jdee! Was there already a flag called `settings`? Sorry, there are so many options for different tools I must have gotten mixed up! If there was already a `settings` we should just go with that! Really sorry for the back and forth!

jdee commented:
No worries, @ohwhtup. I thought that might be the case. I'll revert that commit and submit a PR to the docs repo. Thanks.

jdee committed:
Revert "Also use the new parameter"

This reverts commit 1b4f179f1b17587ed360800096290932d4031cc9.

jdee committed:
Revert "Change config argument to include_settings: to match the env. var."

This reverts commit bddf9e7fe9e1980b5aba69b5d2c891c90df75f07.

ohwutup commented:
Thanks for the PR and working through all the changes @jdee !! 🚀

jdee commented:
My pleasure. Fastlane is a nice tool. I'm glad to be able to contribute.
